cl/corpus_importer/management/commands/import_tn.py

```diff
@@ -0,0 +1,188 @@
+        "judges": panelists,
+        "author_id": lead_author.id,
+        "author": lead_author,
+        "date_filed_is_approximate": False,
+        "blocked_statuses": False,
+        "neutral_citations": case["neutral_citation"],
+        "download_urls": case["pdf_url"],
+    }
+
+
+def add_neutral_citations(filepath):
+    """Add neutral citations to our dataset
+
+    :param tn_corpus: The case data.
+    :type: dict
+    :return: Our case data with neutral citations included.
+    :type: dict
+    """
+    tn_corpus = json.loads(open(filepath, "r").read())
+    results = []
+    for court in ["tennworkcompapp", "tennworkcompcl"]:
+        reporter = "TN WC App." if court == "tennworkcompapp" else "TN WC"
+        tn = [
+            x
+            for x in sorted(
+                tn_corpus, key=lambda x: (x["pub_date"], x["label"])
+            )
+            if x["court"] == court
+        ]
+        for key, group in itertools.groupby(tn, lambda x: x["pub_date"][:4]):
+            count = 1
+            for case in list(group):
+                case["neutral_citation"] = "%s %s %s" % (key, reporter, count)
+                results.append(case)
+                count += 1
+    return results
+
+
+def import_tn_corpus(log, skip_until, dir):
+    """Import TN Corpus
+
+    :param log: Should we view logging info
+    :param skip_until: Label ID, if any, to process first
+    :param dir: Location of our import files
+    :return: None
+    """
+    ready = False if skip_until else True
+
+    if log:
+        logging.getLogger().setLevel(logging.INFO)
+
+    logging.info("Starting import")
+    filepath = "%s/data.json" % dir
+    tn_corpus = add_neutral_citations(filepath)
+    if not ready:
+        case = [x for x in tn_corpus if x["label"] == int(skip_until)][0]
+        logging.info(
+            "Skipping until case %s labeled: %s", case["title"], case["label"]
+        )
+
+    for case in sorted(tn_corpus, key=lambda x: x["label"]):
+        if case["label"] == int(skip_until):
+            ready = True
+        if not ready:
+            continue
+        logging.info(
+            "Processing label:%s for case:%s", case["label"], case["title"]
+        )
+        pdf_path = [
+            x
+            for x in glob("%s/%s/*.pdf" % (dir, case["label"]))
+            if "stamped" not in x
+        ][0]
+        pdf_data = open(pdf_path, "r").read()
+
+        sha1_hash = sha1(force_bytes(pdf_data))
+        ops = Opinion.objects.filter(sha1=sha1_hash)
+        if len(ops) > 0:
+            op = ops[0]
+            logging.warn(
+                "Document already in database. See: %s at %s"
+                % (op.get_absolute_url(), op.cluster.case_name)
+            )
+
+        docket, opinion, cluster, citations, error = make_objects(
+            make_item(case),
+            Court.objects.get(pk=case["court"]),
+            sha1(force_bytes(pdf_data)),
+            open(pdf_path, "r").read(),
```

Comment:
Oof. These three lines all repeat something you already did. You get the Court each and every time instead of caching it, you compute the sha1 a second time, and you open and read the file (without closing it) a second time. Refactor please?

Reply:
I see two of them but not the court object.  I look forward to your next comment on it

Reply:
Yeah, that one will get cached by the DB, so it's not the end of the world, but remember that every time you do Court.objects.get, you send data over the network to the DB, have it look something up and then return it back to you over the network. Instead of doing that thousands of times, you could do it once at the beginning of the script and have the court data ready. 

Something like this near the beginning:

```
courts = {"tennct": Court.objects.get(pk="tennct"), "tennworkcp": Court.objects.get(pk="tennworkcp")}
```

Then on this line you could do:

```
courts[case["court"]]
```

And you'd save the lookup.

Reply:
Gotcha.

Reply:
fixed.